behavior_path_planner crashes when it is launched in .launch.py for the scenario UC-PB-002-0001-Hsinchu

### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I'm convinced that this is not my fault but a bug.

### Description

**This issue is a part of https://github.com/autowarefoundation/autoware.universe/issues/566**

`behavior_path_planner` crashes when it is launched using `behavior_planning.launch.py` which it is running along with other launch files once `UC-PB-002-0001-Hsinchu` is run with `scenario_test_runner` 

```bash
[openscenario_interpreter_node-3] [component_container_mt-13] terminate called after throwing an instance of 'std::bad_alloc'
[openscenario_interpreter_node-3] [component_container_mt-13]   what():  std::bad_alloc
[openscenario_interpreter_node-3] [ERROR] [component_container_mt-13]: process has died [pid 6215, exit code -6, cmd '/opt/ros/galactic/lib/rclcpp_components/component_container_mt --ros-args -r __node:=behavior_planning_container -r __ns:=/planning/scenario_planning/lane_driving/behavior_planning --params-file /tmp/launch_params_mep7xn71'].
```
Full log file is here ([bpp_bad_alloc.txt](https://github.com/autowarefoundation/autoware.universe/files/8845104/bpp_bad_alloc.txt))
Video is [here](https://youtu.be/fqMHIWCp4LA).

Here you can observe that `behavior_planning_container`  is died and it is because `behavior_path_planner` crashes. 

So if one would  launch `scenario_test_runner` after disabling the line `behavior_path_planner_component` [in the launch file](https://github.com/autowarefoundation/autoware.universe/blob/main/launch/tier4_planning_launch/launch/scenario_planning/lane_driving/behavior_planning/behavior_planning.launch.py#:~:text=composable_node_descriptions%3D%5B-,behavior_path_planner_component,-%2C) and then run `behavior_path_planner` with the following code ; it doesn't crash.

```bash
source ~/autoware/install/setup.bash && \
ros2 run behavior_path_planner behavior_path_planner \
--ros-args \
-r __ns:=/planning/scenario_planning/lane_driving/behavior_planning \
-r __node:=behavior_path_planner \
-r behavior_path_planner/input/route:=/planning/mission_planning/route \
-r behavior_path_planner/input/vector_map:=/map/vector_map \
-r behavior_path_planner/input/perception:=/perception/object_recognition/objects \
-r behavior_path_planner/input/odometry:=/localization/kinematic_state \
-r behavior_path_planner/input/scenario:=/planning/scenario_planning/scenario \
-r behavior_path_planner/input/external_approval:=/planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/path_change_approval \
-r behavior_path_planner/input/force_approval:=/planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/path_change_force \
-r behavior_path_planner/output/path:=path_with_lane_id \
-r behavior_path_planner/output/ready:=/planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/ready_module \
-r behavior_path_planner/output/running:=/planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/running_modules \
-r behavior_path_planner/output/force_available:=/planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/force_available \
-r behavior_path_planner/output/turn_indicators_cmd:=/planning/turn_indicators_cmd \
-r behavior_path_planner/output/hazard_lights_cmd:=/planning/hazard_lights_cmd \
-p bt_tree_config_path:=$HOME/autoware/src/universe/autoware.universe/planning/behavior_path_planner/config/behavior_path_planner_tree.xml \
--params-file $HOME/autoware/src/universe/autoware.universe/launch/tier4_planning_launch/config/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/side_shift/side_shift.param.yaml \
--params-file $HOME/autoware/src/universe/autoware.universe/launch/tier4_planning_launch/config/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/avoidance/avoidance.param.yaml \
--params-file $HOME/autoware/src/universe/autoware.universe/launch/tier4_planning_launch/config/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/lane_change/lane_change.param.yaml \
--params-file $HOME/autoware/src/universe/autoware.universe/launch/tier4_planning_launch/config/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/lane_following/lane_following.param.yaml \
--params-file $HOME/autoware/src/universe/autoware.universe/launch/tier4_planning_launch/config/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/pull_over/pull_over.param.yaml \
--params-file $HOME/autoware/src/universe/autoware.universe/launch/tier4_planning_launch/config/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/pull_out/pull_out.param.yaml \
--params-file $HOME/autoware/src/universe/autoware.universe/launch/tier4_planning_launch/config/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/behavior_path_planner.param.yaml \
--params-file $HOME/autoware/src/universe/autoware.universe/vehicle/vehicle_info_util/config/vehicle_info.param.yaml
```

**!!! Please give the right file paths if your autoware repo is now at $HOME !!!** 

### Expected behavior

`behavior_path_planner` shouldn't crash.

### Actual behavior

`behavior_path_planner` crashes for a certain scenario.

### Steps to reproduce

1. Download [map_data](https://gitlab.com/autowarefoundation/operational-design-domains/-/tree/master/public_road_bus/maps/hsinchu).
2. Download the [scenario file](https://gitlab.com/autowarefoundation/operational-design-domains/-/blob/master/public_road_bus/UC-PB-002-0001-Hsinchu.yaml).
3. Launch the `scenario_test_runner` with the following code after setting the right file paths:
```bash
source ~/autoware/install/setup.bash && \
ros2 launch scenario_test_runner scenario_test_runner.launch.py sensor_model:=sample_sensor_kit vehicle_model:=sample_vehicle \
scenario:=$HOME/operational-design-domains/public_road_bus/UC-PB-002-0001-Hsinchu.yaml \
architecture_type:=awf/universe launch_rviz:=true launch_autoware:=true
```


### Versions

- OS: Ubuntu 20.04
- ROS2 Galactic
- https://github.com/autowarefoundation/autoware/commit/5e76cfd1a0359d6c275accadfa5dc2dff9f0d861
- autoware.universe: https://github.com/autowarefoundation/autoware.universe/commit/932ff58be7f51a8461def4bd57876f3bd766e2aa
- https://github.com/tier4/scenario_simulator_v2/commit/7318ece6d59bed4b7a82d648af34113a2e8faf39
- [Operational Design Domains](https://gitlab.com/autowarefoundation/operational-design-domains): [`c589c2`](https://gitlab.com/autowarefoundation/operational-design-domains/-/commit/c589c27fe5768ffb46c8dc8c2243485c99edf9cf) 

### Possible causes

It might be related to the container structure and the memory allocation related to it.

### Additional context

_No response_

--- a/planning/behavior_path_planner/include/behavior_path_planner/scene_module/lane_change/util.hpp
+++ b/planning/behavior_path_planner/include/behavior_path_planner/scene_module/lane_change/util.hpp
@@ -58,14 +58,14 @@ std::vector<LaneChangePath> selectValidPaths(
 bool selectSafePath(
   const std::vector<LaneChangePath> & paths, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects, const Pose & current_pose,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const Pose & current_pose,
   const Twist & current_twist, const double vehicle_width,
   const behavior_path_planner::LaneChangeParameters & ros_parameters,
   LaneChangePath * selected_path);
 bool isLaneChangePathSafe(
   const PathWithLaneId & path, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects, const Pose & current_pose,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const Pose & current_pose,
   const Twist & current_twist, const double vehicle_width,
   const behavior_path_planner::LaneChangeParameters & ros_parameters, const bool use_buffer = true,
   const double acceleration = 0.0);

--- a/planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/util.hpp
+++ b/planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/util.hpp
@@ -66,14 +66,14 @@ std::vector<PullOutPath> selectValidPaths(
 bool selectSafePath(
   const std::vector<PullOutPath> & paths, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects, const Pose & current_pose,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const Pose & current_pose,
   const Twist & current_twist, const double vehicle_width,
   const behavior_path_planner::PullOutParameters & ros_parameters,
   const tier4_autoware_utils::LinearRing2d & vehicle_footprint, PullOutPath * selected_path);
 bool isPullOutPathSafe(
   const behavior_path_planner::PullOutPath & path, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects,
+  const PredictedObjects::ConstSharedPtr dynamic_objects,
   const behavior_path_planner::PullOutParameters & ros_parameters,
   const tier4_autoware_utils::LinearRing2d & vehicle_footprint, const bool use_buffer = true,
   const bool use_dynamic_object = false);

--- a/planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/util.hpp
+++ b/planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/util.hpp
@@ -58,13 +58,13 @@ std::vector<PullOverPath> selectValidPaths(
 bool selectSafePath(
   const std::vector<PullOverPath> & paths, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects, const Pose & current_pose,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const Pose & current_pose,
   const Twist & current_twist, const double vehicle_width,
   const behavior_path_planner::PullOverParameters & ros_parameters, PullOverPath * selected_path);
 bool isPullOverPathSafe(
   const PathWithLaneId & path, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects, const Pose & current_pose,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const Pose & current_pose,
   const Twist & current_twist, const double vehicle_width,
   const behavior_path_planner::PullOverParameters & ros_parameters, const bool use_buffer = true,
   const double acceleration = 0.0);

--- a/planning/behavior_path_planner/src/behavior_path_planner_node.cpp
+++ b/planning/behavior_path_planner/src/behavior_path_planner_node.cpp
@@ -527,10 +527,7 @@ void BehaviorPathPlannerNode::run()
   // update planner data
   planner_data_->self_pose = self_pose_listener_.getCurrentPose();
 
-  // NOTE: planner_data must not be referenced for multithreading
   const auto planner_data = planner_data_;
-  mutex_pd_.unlock();
-
   // run behavior planner
   const auto output = bt_manager_->run(planner_data);
 
@@ -539,7 +536,6 @@ void BehaviorPathPlannerNode::run()
   const auto path_candidate = getPathCandidate(output, planner_data);
 
   // update planner data
-  mutex_pd_.lock();  // for planner_data_
   planner_data_->prev_output_path = path;
   mutex_pd_.unlock();
 

--- a/planning/behavior_path_planner/src/scene_module/lane_change/util.cpp
+++ b/planning/behavior_path_planner/src/scene_module/lane_change/util.cpp
@@ -271,7 +271,7 @@ std::vector<LaneChangePath> selectValidPaths(
 bool selectSafePath(
   const std::vector<LaneChangePath> & paths, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects, const Pose & current_pose,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const Pose & current_pose,
   const Twist & current_twist, const double vehicle_width,
   const LaneChangeParameters & ros_parameters, LaneChangePath * selected_path)
 {
@@ -329,7 +329,7 @@ bool hasEnoughDistance(
 bool isLaneChangePathSafe(
   const PathWithLaneId & path, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects, const Pose & current_pose,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const Pose & current_pose,
   const Twist & current_twist, const double vehicle_width,
   const LaneChangeParameters & ros_parameters, const bool use_buffer, const double acceleration)
 {

--- a/planning/behavior_path_planner/src/scene_module/pull_out/util.cpp
+++ b/planning/behavior_path_planner/src/scene_module/pull_out/util.cpp
@@ -397,8 +397,7 @@ bool hasEnoughDistance(
 bool isPullOutPathSafe(
   const behavior_path_planner::PullOutPath & path, const lanelet::ConstLanelets & road_lanes,
   const lanelet::ConstLanelets & shoulder_lanes,
-  const PredictedObjects::ConstSharedPtr dynamic_objects,
-  const PullOutParameters & ros_parameters,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const PullOutParameters & ros_parameters,
   const tier4_autoware_utils::LinearRing2d & local_vehicle_footprint, const bool use_buffer,
   const bool use_dynamic_object)
 {

--- a/planning/behavior_path_planner/src/scene_module/pull_out/util.cpp
+++ b/planning/behavior_path_planner/src/scene_module/pull_out/util.cpp
@@ -337,7 +337,7 @@ std::vector<PullOutPath> selectValidPaths(
 bool selectSafePath(
   const std::vector<PullOutPath> & paths, const lanelet::ConstLanelets & road_lanes,
   const lanelet::ConstLanelets & shoulder_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects,
+  const PredictedObjects::ConstSharedPtr dynamic_objects,
   [[maybe_unused]] const Pose & current_pose, [[maybe_unused]] const Twist & current_twist,
   [[maybe_unused]] const double vehicle_width, const PullOutParameters & ros_parameters,
   const tier4_autoware_utils::LinearRing2d & local_vehicle_footprint, PullOutPath * selected_path)
@@ -397,7 +397,7 @@ bool hasEnoughDistance(
 bool isPullOutPathSafe(
   const behavior_path_planner::PullOutPath & path, const lanelet::ConstLanelets & road_lanes,
   const lanelet::ConstLanelets & shoulder_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects,
+  const PredictedObjects::ConstSharedPtr dynamic_objects,
   const PullOutParameters & ros_parameters,
   const tier4_autoware_utils::LinearRing2d & local_vehicle_footprint, const bool use_buffer,
   const bool use_dynamic_object)

--- a/planning/behavior_path_planner/src/scene_module/pull_over/util.cpp
+++ b/planning/behavior_path_planner/src/scene_module/pull_over/util.cpp
@@ -305,7 +305,7 @@ std::vector<PullOverPath> selectValidPaths(
 bool selectSafePath(
   const std::vector<PullOverPath> & paths, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects, const Pose & current_pose,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const Pose & current_pose,
   const Twist & current_twist, const double vehicle_width,
   const PullOverParameters & ros_parameters, PullOverPath * selected_path)
 {
@@ -363,7 +363,7 @@ bool hasEnoughDistance(
 bool isPullOverPathSafe(
   const PathWithLaneId & path, const lanelet::ConstLanelets & current_lanes,
   const lanelet::ConstLanelets & target_lanes,
-  const PredictedObjects::ConstSharedPtr & dynamic_objects, const Pose & current_pose,
+  const PredictedObjects::ConstSharedPtr dynamic_objects, const Pose & current_pose,
   const Twist & current_twist, const double vehicle_width,
   const PullOverParameters & ros_parameters, const bool use_buffer, const double acceleration)
 {

--- a/planning/behavior_path_planner/src/utilities.cpp
+++ b/planning/behavior_path_planner/src/utilities.cpp
@@ -653,7 +653,7 @@ std::vector<size_t> filterObjectsByLanelets(
   }
 
   for (size_t i = 0; i < objects.objects.size(); i++) {
-    const auto obj = objects.objects.at(i);
+    const auto & obj = objects.objects.at(i);
     // create object polygon
     Polygon2d obj_polygon;
     if (!calcObjectPolygon(obj, &obj_polygon)) {
@@ -687,7 +687,7 @@ std::vector<size_t> filterObjectsByLanelets(
 
   for (size_t i = 0; i < objects.objects.size(); i++) {
     // create object polygon
-    const auto obj = objects.objects.at(i);
+    const auto & obj = objects.objects.at(i);
     // create object polygon
     Polygon2d obj_polygon;
     if (!calcObjectPolygon(obj, &obj_polygon)) {

